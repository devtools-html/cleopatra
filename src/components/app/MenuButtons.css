/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.menuButtons {
  --internal-share-button-background-color: var(--share-button-background-color, hsl(137,81%,44%));
  --internal-uploading-button-background-color: var(--uploading-button-background-color, #38445f);
  --internal-uploading-progress-fill-color: var(--uploading-progress-fill-color, #7990c8);
  height: 100%;
  display: flex;
  flex-flow: row nowrap;
}

.menuButtonsLink {
  height: 24px;
  position: relative;
  padding: 0 10px;
  border-left: 1px solid var(--grey-30);
  text-decoration: none;
  line-height: 24px;
  cursor: default;
  color: var(--grey-90);
}

.menuButtonsLink:hover {
  background-color: rgba(0, 0, 0, 0.1);
}

.menuButtonsCompositeButtonContainer {
  display: flex;
  flex-flow: column nowrap;
  border-left: 1px solid var(--grey-30);
  border-right: 1px solid var(--grey-30);
}

.menuButtonsShareButton,
.menuButtonsUploadingButton,
.menuButtonsPermalinkButton,
.menuButtonsUploadErrorButton,
.menuButtonsSecondaryShareButton {
  height: 24px;
  margin-bottom: -24px;
}

.menuButtonsUploadingButton {
  overflow: hidden;
}

.menuButtonsUploadingButtonInner {
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
  height: 100%;
  overflow: hidden;
}

.menuButtonsUploadingButtonLabel {
  padding: 0 10px;
  line-height: 24px;
  text-align: center;
  color: hsla(0,0%,100%,0.7);
  position: relative;
  -moz-user-select: none;
  cursor: default;
}

.menuButtonsUploadingButtonProgress {
  height: 24px;
  margin-bottom: -24px;
  width: 100%;
  background: var(--internal-uploading-button-background-color);
  border: 0;
  transform-origin: bottom left;
}

.menuButtonsUploadingButtonProgress::-moz-progress-bar {
  background: var(--internal-uploading-progress-fill-color);
}

.menuButtonsShareButtonButton:not([disabled]),
.menuButtonsSecondaryShareButtonButton:not([disabled]) {
  background-color: #0D8730;
  color: white;
}

.menuButtonsCompositeButtonContainer:not(.currentButtonIsShareButton) > .menuButtonsShareButton,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsUploadingButton) > .menuButtonsUploadingButton,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsPermalinkButton) > .menuButtonsPermalinkButton,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsUploadErrorButton) > .menuButtonsUploadErrorButton,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsSecondaryShareButton) > .menuButtonsSecondaryShareButton {
  pointer-events: none;
}

.menuButtonsShareButtonButton,
.menuButtonsUploadingButtonInner,
.menuButtonsPermalinkButtonButton,
.menuButtonsUploadErrorButtonButton,
.menuButtonsSecondaryShareButtonButton {
  transition: transform 200ms ease-in-out, margin-left 200ms ease-in-out;
}

.menuButtonsCompositeButtonContainer:not(.currentButtonIsShareButton) > .menuButtonsShareButton > .buttonWithPanelButtonWrapper > .menuButtonsShareButtonButton,
.menuButtonsCompositeButtonContainer.currentButtonIsPermalinkButton > .menuButtonsUploadingButton > .menuButtonsUploadingButtonInner,
.menuButtonsCompositeButtonContainer.currentButtonIsUploadErrorButton > .menuButtonsUploadingButton > .menuButtonsUploadingButtonInner,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsSecondaryShareButton) .menuButtonsSecondaryShareButtonButton {
  transform: translateY(-24px);
}

.menuButtonsCompositeButtonContainer.currentButtonIsShareButton > .menuButtonsUploadingButton > .menuButtonsUploadingButtonInner,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsPermalinkButton) > .menuButtonsPermalinkButton > .buttonWithPanelButtonWrapper > .menuButtonsPermalinkButtonButton,
.menuButtonsCompositeButtonContainer:not(.currentButtonIsUploadErrorButton) > .menuButtonsUploadErrorButton > .buttonWithPanelButtonWrapper > .menuButtonsUploadErrorButtonButton {
  transform: translateY(24px);
}

.currentButtonIsPermalinkButton.currentButtonIsSecondaryShareButton .menuButtonsPermalinkButton {
  width: 112px;
}

.currentButtonIsPermalinkButton.currentButtonIsSecondaryShareButton .menuButtonsSecondaryShareButton {
  margin-left: 112px;
}

.menuButtonsPrivacyPanel,
.menuButtonsUploadErrorPanel {
  --width: 280px;
}

.menuButtonsProfileDownloadPanel {
  --offset-from-right: 25px;
  --width: 30em;
}

.menuButtonsDownloadLink {
  white-space: nowrap;
}

.menuButtonsPermalinkTextField {
  width: 14em;
  margin: 5px -5px;
}

.menuButtonsShareNetworkUrlsContainer {
  margin: 20px 0;
  -moz-user-select: none;
  user-select: none;
}

.menuButtonsShareNetworkUrlsCheckbox {
  vertical-align: middle;
  position: relative;
  bottom: 1px;
}

.menuButtonsOpenMetaInfoButtonBox,
.menuButtonsOpenMetaInfoButtonLabel,
.menuButtonsOpenMetaInfoButtonButton {
  height: 100%;
  font-size: 100%;
}

.menuButtonsOpenMetaInfoButtonBox {
  display: flex;
  padding: 0;
  /* Let this box shrink down to the size of the three dots and a bit of text. */
  min-width: 75px;
  /* Shrink down twice as fast as the range selection labels. */
  flex-shrink: 2;
}

.menuButtonsOpenMetaInfoButtonLabel {
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
  position: relative;
  padding: 5px 10px;
  /* Let this box shrink down to the size of the three dots. */
  min-width: 0;
  /* Make sure the padding doesn't contribute to the height, which is 100% */
  box-sizing: border-box;
}

.menuButtonsOpenMetaInfoButtonLabelOverflow {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.menuButtonsOpenMetaInfoButtonBox .buttonWithPanelButton,
.menuButtonsOpenMetaInfoButtonBox .buttonWithPanel:hover,
.menuButtonsOpenMetaInfoButtonBox .buttonWithPanel.open {
  background: url(../../../res/img/svg/meatball.svg) 50% 50% no-repeat;
  width: 40px;
}
