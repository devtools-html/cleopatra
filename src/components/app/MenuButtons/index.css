/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.menuButtonsButton {
  display: flex;
  height: 100%;
  align-items: center;
  padding: 0 20px;
  border: 0;
  border-left: 1px solid var(--grey-30);
  margin: 0;
  background: none;

  /* Inherit the color rather than using the browser's */
  color: inherit;
  font-size: 100%;
  white-space: nowrap;
}

.menuButtonsButton:hover {
  background-color: rgba(0, 0, 0, 0.1);
}

.buttonWithPanel.open > .menuButtonsButton,
.menuButtonsButton:hover:active {
  background-color: rgba(0, 0, 0, 0.2);
}

.menuButtonsLink {
  position: relative; /* This allows positioning the ::before element. */
  cursor: default;
  text-decoration: none;
}

.open-in-new {
  padding: 10px;
  background-image: url(firefox-profiler-res/img/svg/open-in-new.svg);
  background-position: center;
  background-repeat: no-repeat;
  background-size: 11px;
}
