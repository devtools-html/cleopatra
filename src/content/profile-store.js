

export function uploadBinaryProfileData(data, progressChangeCallback = undefined) {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();

    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(xhr.responseText);
      } else {
        reject(`xhr onload with status != 200, xhr.statusText: ${xhr.statusText}`);
      }
    };

    xhr.onerror = () => {
      reject(`xhr onerror was called, xhr.statusText: ${xhr.statusText}`);
    };

    xhr.upload.onprogress = e => {
      if (progressChangeCallback && e.lengthComputable) {
        progressChangeCallback(e.loaded / e.total);
      }
    };

    xhr.open('POST', 'https://profile-store.appspot.com/compressed-store');
    xhr.send(data);
  });
}
