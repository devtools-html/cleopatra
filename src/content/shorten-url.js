import fetchJsonP from 'fetch-jsonp';
import queryString from 'query-string';
import url from 'url';

export default function shortenURL(urlToShorten) {
  let longURL = urlToShorten;
  if (!longURL.startsWith('https://perf-html.io/')) {
    const parsedURL = url.parse(longURL);
    const parsedURLOnCanonicalHost = Object.assign({}, parsedURL, {
      protocol: 'https:',
      host: 'perf-html.io',
    });
    longURL = url.format(parsedURLOnCanonicalHost);
  }

  const bitlyQueryURL = 'https://api-ssl.bitly.com/v3/shorten?' +
    queryString.stringify({
      'longUrl': longURL,
      'domain': 'perfht.ml',
      'access_token': 'b177b00a130faf3ecda6960e8b59fde73e902422',
    });
  return fetchJsonP(bitlyQueryURL).then(response => response.json()).then(json => json.data.url);
}
