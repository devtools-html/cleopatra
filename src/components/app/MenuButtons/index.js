/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import explicitConnect from '../../../utils/connect';
import { getProfile, getProfileRootRange } from '../../../selectors/profile';
import {
  getDataSource,
  getIsNewlyPublished,
} from '../../../selectors/url-state';
import { MenuButtonsMetaInfo } from './MetaInfo';
import { MenuButtonsPublish } from './Publish';
import { MenuButtonsPermalink } from './Permalink';
import { assertExhaustiveCheck } from '../../../utils/flow';
import ArrowPanel from '../../shared/ArrowPanel';
import ButtonWithPanel from '../../shared/ButtonWithPanel';
import { dismissNewlyPublished } from '../../../actions/app';

import type { StartEndRange } from '../../../types/units';
import type { Profile } from '../../../types/profile';
import type { DataSource } from '../../../types/actions';
import type { ConnectedProps } from '../../../utils/connect';

require('./index.css');

type OwnProps = {|
  // This is for injecting a URL shortener for tests. Normally we would use a Jest mock
  // that would mock out a local module, but I was having trouble getting it working
  // correctly (perhaps due to ES6 modules), so I just went with dependency injection
  // instead.
  injectedUrlShortener?: string => Promise<string>,
|};

type StateProps = {|
  +profile: Profile,
  +rootRange: StartEndRange,
  +dataSource: DataSource,
  +isNewlyPublished: boolean,
|};

type DispatchProps = {|
  +dismissNewlyPublished: typeof dismissNewlyPublished,
|};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

class MenuButtons extends React.PureComponent<Props> {
  componentDidMount() {
    // Clear out the newly published notice from the URL.
    this.props.dismissNewlyPublished();
  }

  render() {
    const {
      profile,
      dataSource,
      isNewlyPublished,
      injectedUrlShortener,
    } = this.props;
    return (
      <>
        {/* Place the info button outside of the menu buttons to allow it to shrink. */}
        <MenuButtonsMetaInfo profile={profile} />
        <div className="menuButtons">
          <PublishOrPermalinkButtons
            dataSource={dataSource}
            isNewlyPublished={isNewlyPublished}
            injectedUrlShortener={injectedUrlShortener}
          />
          <a
            href="/docs/"
            target="_blank"
            className="menuButtonsLink"
            title="Open the documentation in a new window"
          >
            Docs
            <i className="open-in-new" />
          </a>
        </div>
      </>
    );
  }
}

const PublishOrPermalinkButtons = ({
  dataSource,
  isNewlyPublished,
  injectedUrlShortener,
}) => {
  switch (dataSource) {
    case 'from-addon':
    case 'from-file':
    case 'local':
      return (
        <ButtonWithPanel
          className="menuButtonsShareButton"
          label="Publishâ€¦"
          panel={
            <ArrowPanel className="menuButtonsPublishPanel">
              <MenuButtonsPublish />
            </ArrowPanel>
          }
        />
      );
    case 'public':
    case 'from-url':
    case 'compare':
      return (
        <MenuButtonsPermalink
          isNewlyPublished={isNewlyPublished}
          injectedUrlShortener={injectedUrlShortener}
        />
      );
    case 'none':
      return null;
    default:
      throw assertExhaustiveCheck(dataSource);
  }
};

export default explicitConnect<OwnProps, StateProps, DispatchProps>({
  mapStateToProps: state => ({
    profile: getProfile(state),
    rootRange: getProfileRootRange(state),
    dataSource: getDataSource(state),
    isNewlyPublished: getIsNewlyPublished(state),
  }),
  mapDispatchToProps: {
    dismissNewlyPublished,
  },
  component: MenuButtons,
});
