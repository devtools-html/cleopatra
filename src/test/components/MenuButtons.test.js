/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import MenuButtons from '../../components/app/MenuButtons';
import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { TextEncoder } from 'util';
import { stateFromLocation } from '../../app-logic/url-handling';
import { ensureExists } from '../../utils/flow';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';

// Mocking SymbolStoreDB
import { uploadBinaryProfileData } from '../../profile-logic/profile-store';
jest.mock('../../profile-logic/profile-store');

// Mocking sha1
import sha1 from '../../utils/sha1';
jest.mock('../../utils/sha1');

// Mocking compress
jest.mock('../../utils/gz');

// Mocking shortenUrl
import { shortenUrl } from '../../utils/shorten-url';
jest.mock('../../utils/shorten-url');

// Mock hash
const hash = 'c5e53f9ab6aecef926d4be68c84f2de550e2ac2f';

describe('app/MenuButtons', function() {
  function mockUpload() {
    // Create a promise with the resolve function outside of it.
    let resolveUpload, rejectUpload;
    const promise = new Promise((resolve, reject) => {
      resolveUpload = resolve;
      rejectUpload = reject;
    });

    // Flow doesn't know uploadBinaryProfileData is a jest mock.
    (uploadBinaryProfileData: any).mockImplementation(
      (() => ({
        abortFunction: () => {},
        startUpload: () => promise,
      }): typeof uploadBinaryProfileData)
    );

    return { resolveUpload, rejectUpload };
  }

  function setup(updateChannel = 'release') {
    const { profile } = getProfileFromTextSamples('A');
    profile.meta.updateChannel = updateChannel;
    const store = storeWithProfile(profile);
    const { resolveUpload, rejectUpload } = mockUpload();

    store.dispatch({
      type: 'UPDATE_URL_STATE',
      newUrlState: stateFromLocation({
        pathname: '/from-addon',
        search: '',
        hash: '',
      }),
    });

    const renderResult = render(
      <Provider store={store}>
        <MenuButtons />
      </Provider>
    );

    const { container, getByTestId, getByText } = renderResult;
    const getPublishButton = () => getByText('Publishâ€¦');
    const getPanelForm = () =>
      ensureExists(
        container.querySelector('form'),
        'Could not find the form in the panel'
      );
    const getPanel = () => getByTestId('MenuButtonsPublish-container');

    return {
      store,
      ...renderResult,
      getPanel,
      getPublishButton,
      getPanelForm,
      resolveUpload,
      rejectUpload,
    };
  }

  describe('<Publish>', function() {
    beforeAll(function() {
      if ((window: any).TextEncoder) {
        throw new Error('A TextEncoder was already on the window object.');
      }
      (window: any).TextEncoder = TextEncoder;
    });

    afterAll(async function() {
      delete URL.createObjectURL;
      delete URL.revokeObjectURL;
      delete (window: any).TextEncoder;
    });

    beforeEach(function() {
      // Flow doesn't know sha1 is a jest mock.
      (sha1: any).mockImplementation((_data: Uint8Array) =>
        Promise.resolve(hash)
      );
      // Flow doesn't know shortenUrl is a jest mock.
      (shortenUrl: any).mockImplementation(() =>
        Promise.resolve('https://profiler.firefox.com/')
      );
      // jsdom does not have URL.createObjectURL.
      // See https://github.com/jsdom/jsdom/issues/1721
      (URL: any).createObjectURL = () => 'mockCreateObjectUrl';
      (URL: any).revokeObjectURL = () => {};
    });

    it('matches the snapshot for the closed state', () => {
      const { container } = setup();
      expect(container).toMatchSnapshot();
    });

    it('matches the snapshot for the opened panel for a nightly profile', () => {
      const { getPanel, getPublishButton } = setup('nightly');
      fireEvent.click(getPublishButton());
      expect(getPanel()).toMatchSnapshot();
    });

    it('matches the snapshot for the opened panel for a release profile', () => {
      const { getPanel, getPublishButton } = setup('release');
      fireEvent.click(getPublishButton());
      expect(getPanel()).toMatchSnapshot();
    });

    it('matches the snapshot for the uploading panel', () => {
      const { getPanel, getPublishButton, getPanelForm } = setup();
      fireEvent.click(getPublishButton());
      fireEvent.submit(getPanelForm());
      expect(getPanel()).toMatchSnapshot();
    });

    it('matches the snapshot for the completed upload panel', () => {
      const {
        getPanel,
        getPublishButton,
        getPanelForm,
        resolveUpload,
      } = setup();
      fireEvent.click(getPublishButton());
      fireEvent.submit(getPanelForm());
      resolveUpload();
      expect(getPanel()).toMatchSnapshot();
    });

    it('matches the snapshot for an error', () => {
      const {
        getPanel,
        getPublishButton,
        getPanelForm,
        rejectUpload,
      } = setup();
      fireEvent.click(getPublishButton());
      fireEvent.submit(getPanelForm());
      rejectUpload('This is a mock error');
      expect(getPanel()).toMatchSnapshot();
    });
  });
});
