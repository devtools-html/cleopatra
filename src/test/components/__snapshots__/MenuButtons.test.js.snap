// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/MenuButtons <MenuButtonsMetaInfo> matches the snapshot 1`] = `
<div
  class="buttonWithPanel menuButtonsMetaInfoButton open"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton menuButtonsMetaInfoButtonButton"
      type="button"
      value="Firefox (48.0) Intel Mac OS X 10.11"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open arrowPanelOpenMetaInfo"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <h2
          class="arrowPanelSubTitle"
        >
          Timing
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Recording started:
            </span>
            toLocaleString Sat, 09 Apr 2016 17:02:32 GMT
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Interval:
            </span>
            1
            ms
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Profile Version:
            </span>
            24
          </div>
        </div>
        <h2
          class="arrowPanelSubTitle"
        >
          Application
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Name:
            </span>
            Firefox
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Version:
            </span>
            48.0
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Update Channel:
            </span>
            nightly
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Build ID:
            </span>
            20181126165837
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Build Type:
            </span>
            Debug
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Extensions:
            </span>
            <ul
              class="metaInfoList"
            >
              <li
                class="metaInfoListItem"
              >
                Form Autofill
              </li>
              <li
                class="metaInfoListItem"
              >
                Firefox Screenshots
              </li>
              <li
                class="metaInfoListItem"
              >
                Gecko Profiler
              </li>
            </ul>
          </div>
        </div>
        <h2
          class="arrowPanelSubTitle"
        >
          Platform
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Platform:
            </span>
            Macintosh
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              OS:
            </span>
            Intel Mac OS X 10.11
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              ABI:
            </span>
            x86_64-gcc3
          </div>
        </div>
        <details>
          <summary
            class="arrowPanelSubTitle"
          >
            Profiler Overhead
          </summary>
          <div
            class="arrowPanelSection"
          >
            <div
              class="metaInfoGrid"
            >
              <div />
              <div>
                Mean
              </div>
              <div>
                Max
              </div>
              <div>
                Min
              </div>
              <div>
                Overhead
              </div>
              <div>
                227ns
              </div>
              <div>
                410ns
              </div>
              <div>
                38.0ns
              </div>
              <div>
                Cleaning
              </div>
              <div>
                54.1ns
              </div>
              <div>
                100ns
              </div>
              <div>
                7.00ns
              </div>
              <div>
                Counter
              </div>
              <div>
                59.1ns
              </div>
              <div>
                105ns
              </div>
              <div>
                12.0ns
              </div>
              <div>
                Interval
              </div>
              <div>
                857ns
              </div>
              <div>
                1,000ns
              </div>
              <div>
                0.0000ns
              </div>
              <div>
                Lockings
              </div>
              <div>
                49.1ns
              </div>
              <div>
                95.0ns
              </div>
              <div>
                2.00ns
              </div>
            </div>
            <div
              class="metaInfoRow"
            >
              <span
                class="metaInfoWideLabel"
              >
                Overhead Durations:
              </span>
              <span
                class="metaInfoValueRight"
              >
                1,586ns
              </span>
            </div>
            <div
              class="metaInfoRow"
            >
              <span
                class="metaInfoWideLabel"
              >
                Overhead Percentage:
              </span>
              <span
                class="metaInfoValueRight"
              >
                26,433%
              </span>
            </div>
            <div
              class="metaInfoRow"
            >
              <span
                class="metaInfoWideLabel"
              >
                Profiled Duration:
              </span>
              <span
                class="metaInfoValueRight"
              >
                6.00ns
              </span>
            </div>
          </div>
        </details>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/MenuButtons <MenuButtonsMetaInfo> with no statistics object should not make the app crash 1`] = `
<div
  class="buttonWithPanel menuButtonsMetaInfoButton open"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton menuButtonsMetaInfoButtonButton"
      type="button"
      value="Firefox (48.0) Intel Mac OS X 10.11"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open arrowPanelOpenMetaInfo"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <h2
          class="arrowPanelSubTitle"
        >
          Timing
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Recording started:
            </span>
            toLocaleString Sat, 09 Apr 2016 17:02:32 GMT
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Interval:
            </span>
            1
            ms
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Profile Version:
            </span>
            24
          </div>
        </div>
        <h2
          class="arrowPanelSubTitle"
        >
          Application
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Name:
            </span>
            Firefox
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Version:
            </span>
            48.0
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Update Channel:
            </span>
            nightly
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Build ID:
            </span>
            20181126165837
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Build Type:
            </span>
            Debug
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Extensions:
            </span>
            <ul
              class="metaInfoList"
            >
              <li
                class="metaInfoListItem"
              >
                Form Autofill
              </li>
              <li
                class="metaInfoListItem"
              >
                Firefox Screenshots
              </li>
              <li
                class="metaInfoListItem"
              >
                Gecko Profiler
              </li>
            </ul>
          </div>
        </div>
        <h2
          class="arrowPanelSubTitle"
        >
          Platform
        </h2>
        <div
          class="arrowPanelSection"
        >
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              Platform:
            </span>
            Macintosh
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              OS:
            </span>
            Intel Mac OS X 10.11
          </div>
          <div
            class="metaInfoRow"
          >
            <span
              class="metaInfoLabel"
            >
              ABI:
            </span>
            x86_64-gcc3
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for an error 1`] = `
<div
  class="menuButtonsPublishUpload"
  data-testid="MenuButtonsPublish-container"
>
  <div
    class="photon-message-bar photon-message-bar-error"
  >
    Uh oh, something went wrong when publishing the profile.
    <button
      class="photon-button photon-button-micro"
      type="button"
    >
      Try again
    </button>
  </div>
  <div
    class="menuButtonsPublishError"
  >
    This is a mock error
  </div>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the closed state 1`] = `
<div>
  <div
    class="buttonWithPanel menuButtonsMetaInfoButton"
  >
    <div
      class="buttonWithPanelButtonWrapper"
    >
      <input
        class="buttonWithPanelButton menuButtonsMetaInfoButtonButton"
        type="button"
        value="Firefox () "
      />
    </div>
    <div
      class="arrowPanelAnchor"
    >
      <div
        class="arrowPanel arrowPanelOpenMetaInfo"
      >
        <div
          class="arrowPanelArrow"
        />
      </div>
    </div>
  </div>
  <div
    class="menuButtons"
  >
    <div
      class="buttonWithPanel menuButtonsShareButton menuButtonsShareButtonOriginal"
    >
      <div
        class="buttonWithPanelButtonWrapper"
      >
        <input
          class="buttonWithPanelButton"
          type="button"
          value="Publish…"
        />
      </div>
      <div
        class="arrowPanelAnchor"
      >
        <div
          class="arrowPanel menuButtonsPublishPanel"
        >
          <div
            class="arrowPanelArrow"
          />
        </div>
      </div>
    </div>
    <a
      class="menuButtonsLink"
      href="/docs/"
      target="_blank"
      title="Open the documentation in a new window"
    >
      Docs
      <i
        class="open-in-new"
      />
    </a>
  </div>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the opened panel for a nightly profile 1`] = `
<div
  data-testid="MenuButtonsPublish-container"
>
  <form
    class="menuButtonsPublishContent"
  >
    <div
      class="menuButtonsPublishIcon"
    />
    <h1
      class="menuButtonsPublishTitle"
    >
      Share Performance Profile
    </h1>
    <p
      class="menuButtonsPublishInfoDescription"
    >
      Upload your profile and make it accessible to anyone with the link.
       This profile is from Firefox Nightly, so by default all information is included.
    </p>
    <h3>
      Include additional data that may be identifiable
    </h3>
    <div
      class="menuButtonsPublishDataChoices"
    >
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeHiddenThreads"
          type="checkbox"
        />
        Include hidden threads
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeFullTimeRange"
          type="checkbox"
        />
        Include hidden time range
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeScreenshots"
          type="checkbox"
        />
        Include screenshots
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeUrls"
          type="checkbox"
        />
        Include resource URLs
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          checked=""
          class="photon-checkbox photon-checkbox-default"
          name="includeExtension"
          type="checkbox"
        />
        Include extension information
      </label>
    </div>
    <div
      class="menuButtonsPublishButtons"
    >
      <button
        class="photon-button menuButtonsPublishButton menuButtonsPublishButtonsDownload menuButtonsPublishButtonDisabled"
        type="button"
      >
        Compressing…
      </button>
      <button
        class="photon-button photon-button-primary menuButtonsPublishButton menuButtonsPublishButtonsUpload"
        type="submit"
      >
        <span
          class="menuButtonsPublishButtonsSvg menuButtonsPublishButtonsSvgUpload"
        />
        Publish
      </button>
    </div>
  </form>
</div>
`;

exports[`app/MenuButtons <Publish> matches the snapshot for the opened panel for a release profile 1`] = `
<div
  data-testid="MenuButtonsPublish-container"
>
  <form
    class="menuButtonsPublishContent"
  >
    <div
      class="menuButtonsPublishIcon"
    />
    <h1
      class="menuButtonsPublishTitle"
    >
      Share Performance Profile
    </h1>
    <p
      class="menuButtonsPublishInfoDescription"
    >
      Upload your profile and make it accessible to anyone with the link.
       By default, your personal data is removed.
    </p>
    <h3>
      Include additional data that may be identifiable
    </h3>
    <div
      class="menuButtonsPublishDataChoices"
    >
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeHiddenThreads"
          type="checkbox"
        />
        Include hidden threads
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeFullTimeRange"
          type="checkbox"
        />
        Include hidden time range
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeScreenshots"
          type="checkbox"
        />
        Include screenshots
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeUrls"
          type="checkbox"
        />
        Include resource URLs
      </label>
      <label
        class="photon-label menuButtonsPublishDataChoicesLabel"
      >
        <input
          class="photon-checkbox photon-checkbox-default"
          name="includeExtension"
          type="checkbox"
        />
        Include extension information
      </label>
    </div>
    <div
      class="menuButtonsPublishButtons"
    >
      <button
        class="photon-button menuButtonsPublishButton menuButtonsPublishButtonsDownload menuButtonsPublishButtonDisabled"
        type="button"
      >
        Compressing…
      </button>
      <button
        class="photon-button photon-button-primary menuButtonsPublishButton menuButtonsPublishButtonsUpload"
        type="submit"
      >
        <span
          class="menuButtonsPublishButtonsSvg menuButtonsPublishButtonsSvgUpload"
        />
        Publish
      </button>
    </div>
  </form>
</div>
`;
